### PR TITLE
Feature/add ability to skip records

### PR DESCRIPTION
**Pull request by ikbalkaya**
Add ability to skip messages when as described https://github.com/ably/kafka-connect-ably/issues/85
Closes https://github.com/ably/kafka-connect-ably/issues/85

### CODE REVIEW

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java:71]: Too wide variable scope: As far as I can see `messageConfig` and `channelConfig` are only used if `skipOnKeyAbsence` is `true`, so I would move these two `connectorConfig.getString()` calls to within the check below.

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java:74]: Horizontal spacing:

```suggestion
        if (skipOnKeyAbsence) {
```

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java:79]: The split and subsequent need for string concatenation here seems a bit arbitrary and unnecessary. I would just put this whole statement on one line. 🤷

File: src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java:180]: Feels like unnecessary string split and concatenation.

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:13]: Horizontal spacing:

```suggestion
    public static class Result {
```

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:28]: I would add a blank line below this line for readability.

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:46]: Horizontal spacing:

```suggestion
            return new Result(false, null);
```

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:56]: Horizontal spacing:

```suggestion
            if (skippable) {
```

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:63]: Horizontal spacing:

```suggestion
            return new Result(false, pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic()));
```

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:14]: Could this field, and `value` below it, be `private`?

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:38]: This sort of Git noise, being a line unrelated to the change you're making being changed, is why I avoid attempting to line up 'table layouts' in source code. A single space after the param names is all that's needed.

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java:70]: Horizontal spacing:

```suggestion
        // check both channel and message config for skipping the record
```

(though I do recognise there's an earlier comment in this same file where there isn't a space after the `//`, however I do think it's fairly conventional - for readability - to have a space here)

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
The implementation looks sensible, though I've really not got enough depth of knowledge in this domain to say for sure.

I recognise we're yet to work on https://github.com/ably/kafka-connect-ably/issues/74, however I still couldn't resist a lint-like comments on things that bother me when scanning the code by eye.

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java]: Good point - I also moved this to a new method (for now) as this method was getting bigger. Change : 02939fd5e7f6ec18954639a8d1a6e687c3393212

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java]: This split is based on line length set on IDE. You can see this elsewhere in this codebase, for example `ChannelSinkConnectorConfig` will have similar splits.

File: src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java:180]: As above

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java:28]: 110e0282f67f57de0aa6845cb9e3ddd666a1ad1a

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java]: Of course. I also made them `final` as there was no reason for them to be mutable e99f2580953029950c3c37284d7e1aed97d926a4

File: src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/config/ConfigValueEvaluator.java]: 1726377c47596d0033c5dd4728e8c7b880d372dd

File: src/test/java/com/ably/kafka/connect/ConfigValueEvaluatorTest.java
Comment (KacperKluka) [inline src/test/java/com/ably/kafka/connect/ConfigValueEvaluatorTest.java:125]: Is it some kind of TODO comment or does it describe the below test case? If the latter maybe we should put this information in the test name rather than in a comment above it? 🤔

**Review by KacperKluka — COMMENTED** [review]
Overall looks good to me 👍 Due to being unexperienced with Kafka's architecture, I'd prefer to leave my review as a comment and wait for approval from someone more experienced in this technology 😉

File: src/test/java/com/ably/kafka/connect/ConfigValueEvaluatorTest.java
Comment (ikbalkaya) [reply src/test/java/com/ably/kafka/connect/ConfigValueEvaluatorTest.java:125]: It describes the test case indeed

File: src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java
Comment (QuintinWillison) [reply src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java]: I'm increasingly of the opinion that any line length limit is arbitrary and will always be subjective. Is your IDE instructed by any configuration in this codebase? Or is this your own local configuration and/or IDE default? Would exceeding this limit have failed in CI checks?

My preference would be for you to circumvent or otherwise disable this check so that the code can be laid out logically.

Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java]: > Or is this your own local configuration and/or IDE default?


.editorconfig is set on this project. But there is not any enforced line length in it, So editor shows its default line length. Adding max line length setting will just change the line the horizontal spacing from start line and will not effect re-format. Please see:
Screenshot 2022-10-31 at 17 47 23

So the answer is that this is not automatic and there is no CI checks on this repo which will fail if there is an issue with formatting, but also not completely arbitrary too. As I commented on the top, a similar line length has been followed throughout this codebase. Can we keep this like this now until we set up a lint checker on our CI?

Comment (QuintinWillison) [reply src/main/java/com/ably/kafka/connect/client/DefaultAblyClient.java]: Sure and thanks for explaining. 😁 ... I'm happy to leave as is for now and address later on when we come to setting up check style or other lint tool.

File: src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java
Comment (QuintinWillison) [reply src/main/java/com/ably/kafka/connect/config/ChannelSinkConnectorConfig.java:180]: https://github.com/ably/kafka-connect-ably/pull/89#discussion_r1010313991

**Review by QuintinWillison — APPROVED** [review]
I don't have Kafka experience for my approval to carry weight in respect of validity and suitability in that ecosystem, however the code looks solid enough to me. Thanks.

**Pull request merged**
